File: src/domain/argocd.py
"""ArgoCD integration domain model."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping, Optional


@dataclass(frozen=True)
class ArgocdConfig:
    enabled: bool
    namespace: str = "argocd"
    node_port: Optional[int] = None
    plaintext: bool = False

    @classmethod
    def disabled(cls) -> "ArgocdConfig":
        return cls(enabled=False)

    @classmethod
    def from_host_config(
        cls,
        host_config: Mapping[str, Any],
        group_vars: Optional[Mapping[str, Any]] = None,
    ) -> "ArgocdConfig":
        def _get(key: str, default: Any = None) -> Any:
            val = host_config.get(key)
            if val is None and group_vars is not None:
                val = group_vars.get(key)
            return val if val is not None else default

        enabled = bool(_get("argocd_enabled", False))
        if not enabled:
            return cls.disabled()

        namespace = str(_get("argocd_namespace", "argocd"))
        node_port_raw = _get("argocd_node_port")
        node_port = int(node_port_raw) if node_port_raw is not None else None
        plaintext = bool(_get("argocd_plaintext", False))

        return cls(enabled=True, namespace=namespace, node_port=node_port, plaintext=plaintext)

File: internal/domain/argocd.go
package domain

// ArgocdConfig holds ArgoCD integration settings for a cluster host.
type ArgocdConfig struct {
	Enabled   bool
	Namespace string
	NodePort  *int
	Plaintext bool
}

// DisabledArgocdConfig returns a config with ArgoCD disabled.
func DisabledArgocdConfig() ArgocdConfig {
	return ArgocdConfig{Enabled: false, Namespace: "argocd"}
}

// EnabledArgocdConfig returns a minimal enabled config with namespace defaults applied.
func EnabledArgocdConfig() ArgocdConfig {
	return ArgocdConfig{Enabled: true, Namespace: "argocd"}
}

// ArgocdConfigFromHostConfig builds an ArgocdConfig from Ansible inventory maps.
// host_config takes precedence; group_vars is the fallback.
func ArgocdConfigFromHostConfig(hostConfig, groupVars map[string]any) ArgocdConfig {
	get := func(key string) (any, bool) {
		if hostConfig != nil {
			if v, ok := hostConfig[key]; ok {
				return v, true
			}
		}
		if groupVars != nil {
			if v, ok := groupVars[key]; ok {
				return v, true
			}
		}
		return nil, false
	}

	enabledRaw, _ := get("argocd_enabled")
	if !toBool(enabledRaw) {
		return DisabledArgocdConfig()
	}

	ns := "argocd"
	if nsRaw, ok := get("argocd_namespace"); ok {
		if s, ok := nsRaw.(string); ok && s != "" {
			ns = s
		}
	}

	var nodePort *int
	if npRaw, ok := get("argocd_node_port"); ok && npRaw != nil {
		if p := toInt(npRaw); p != 0 {
			nodePort = &p
		}
	}

	plaintext := false
	if ptRaw, ok := get("argocd_plaintext"); ok {
		plaintext = toBool(ptRaw)
	}

	return ArgocdConfig{
		Enabled:   true,
		Namespace: ns,
		NodePort:  nodePort,
		Plaintext: plaintext,
	}
}

func toBool(v any) bool {
	if v == nil {
		return false
	}
	switch val := v.(type) {
	case bool:
		return val
	case int:
		return val != 0
	}
	return false
}

func toInt(v any) int {
	switch val := v.(type) {
	case int:
		return val
	case int64:
		return int(val)
	case float64:
		return int(val)
	}
	return 0
}
